allennlp/training/metrics/bleu.py

```diff
@@ -134,7 +135,29 @@ def __call__(
             self._reference_lengths += valid_gold_targets_mask.sum().item()
 
     @overrides
-    def get_metric(self, reset: bool = False) -> Dict[str, float]:
+    def get_metric(
+        self,
+        reset: bool = False,
+        world_size: int = 1,
+        cuda_device: Union[int, torch.device] = torch.device("cpu"),
+    ) -> Dict[str, float]:
+
+        if world_size > 1:
+            _prediction_lengths = torch.tensor(self._prediction_lengths).to(cuda_device)
+            _reference_lengths = torch.tensor(self._reference_lengths).to(cuda_device)
+            for ngram in self._precision_matches:
```

Comment:
Concerned about efficiency here.

Reply:
I could be wrong, but I don't think this will produce the right results.

`self._precision_matches` could have different keys (ngram indices) for each worker, right? So each time `all_reduce` is used, it might be reducing on mismatched ngrams

Reply:
The fact that this works at all makes it clear to me that I don't understand how `torch.dist` works. If one worker has seen 10 different ngrams, and another has seen 11, wouldn't this hang? How could it possibly match up the tensors from the worker process in the `all_reduce` call, let alone match them to the correct ngram (like @epwalsh already noted)?

Does `torch.dist` only work if all the workers take the exact same code path? That seems like an insanely dangerous programming model.

Reply:
The way I understand it, anytime a node calls any sort of distributed syncronization/communication function (e.g. a reduce function, a barrier, etc.), that call will block until it can communicate with all "matching" calls from the other nodes. The other nodes don't need to take the same code path, but they do need to call the same function.

So if there's a mismatched number of reduce operation calls between nodes, I think the nodes with the most reduce calls would essentially freeze up indefinitely as they wait for a matching call from the other nodes, right?

Reply:
I've moved the `all_reduce` now. Since `_ngram_weights` will be the same for all workers, the number of reduce calls should be the same.